src/unstract/sdk/index.py

```diff
@@ -343,7 +323,18 @@ def generate_file_id(
         Returns:
             str: Key representing unique ID for a file
         """
-        return (
-            f"{tool_id}|{vector_db}|{embedding}|{x2text}|"
-            f"{chunk_size}|{chunk_overlap}|{file_hash}"
-        )
+        index_key = {
+            "tool_id": tool_id,
+            "file_hash": file_hash,
+            "vector_db_config": ToolAdapter.get_adapter_config(
+                self.tool, vector_db
+            ),
+            "embedding_config": ToolAdapter.get_adapter_config(
+                self.tool, embedding
+            ),
+            "x2text_config": ToolAdapter.get_adapter_config(self.tool, x2text),
+            "chunk_size": chunk_size,
+            "chunk_overlap": chunk_overlap,
+        }
+        hashed_index_key = ToolUtils.hash_str(ToolUtils.json_to_str(index_key))
```

Comment:
@chandrasekharan-zipstack Couple of things:

1. JSON spec as such does not enforce key order, and is based on language implementation. Thus we cannot assume hashing a stringified JSON is safe, as the order of the contents may not be reliable.

2. We need to bring in file hash generation logic also inside this doc ID generator, as it is a logical part of doc ID itself and not used anywhere else.

Reply:
@hari-kuriakose I can implement 1 - we'll be safe to future changes that way.
Regarding 2, I'll not bring a major breaking change to the args since its used in other contexts as well. I'll make the `file_hash` an optional arg along with `file_path`.
I think we added this `file_hash` argument to `index_file()` since we'd already compute this during workflow runs with an intention to reuse the same